overlord/snapshotstate/backend/backend.go

```diff
@@ -445,6 +445,119 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 	return nil
 }
 
+// Import a snapshot from the export file format
+func Import(ctx context.Context, id uint64, r io.Reader) (int64, []string, error) {
+	comment := fmt.Sprintf("snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	p := path.Join(dirs.SnapCacheDir, "snapshots", fmt.Sprintf("import-%d", id))
+	if err := os.MkdirAll(path.Join(dirs.SnapCacheDir, "snapshots"), 0755); err != nil {
+		return 0, nil, fmt.Errorf("cannot create snapshots directory")
+	}
+	if err := os.Mkdir(p, 0755); err != nil {
+		if os.IsExist(err) {
+			return 0, nil, fmt.Errorf("snapshot import already in progress for ID `%d`", id)
+		}
+		return 0, nil, fmt.Errorf("failed creating import cache: %v", err)
+	}
+	defer os.RemoveAll(p)
+
+	// unpack the tar file to a temporary location (so the contents can be validated)
+	exportFound, size, err := unpackSnapshotImport(r, p)
+	if err != nil {
+		return 0, nil, err
+	}
+	if !exportFound {
+		return 0, nil, fmt.Errorf("snapshot import file incomplete: no export.json file")
+	}
+
+	// XXX: check the snapshot hashes (these are not in export.json at present)
+
+	// walk the cache directory to store the files
+	dir, err := osOpen(p)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed opening import cache: %v", comment)
+	}
+	defer dir.Close()
+	names, err := dirNames(dir, 100)
+	if err != nil {
+		return 0, nil, fmt.Errorf("failed read from import cache: %v", comment)
+	}
+
+	// move the files into place with the new local set ID
+	snapNames, err := moveCachedSnapshots(names, id, p)
+	return size, snapNames, err
+}
+
+func unpackSnapshotImport(r io.Reader, p string) (exportFound bool, size int64, err error) {
+	tr := tar.NewReader(r)
+	var tarErr error
+	var header *tar.Header
+	for tarErr == nil {
+		var skip bool
+		header, tarErr = tr.Next()
+		switch {
+		case tarErr == io.EOF:
+			skip = true
+		case tarErr != nil:
+			return false, 0, fmt.Errorf("failed reading snapshot import: %v", tarErr)
+		case header == nil:
+			// should not happen
+			skip = true
+		case header.Typeflag == tar.TypeDir:
+			return false, 0, errors.New("unexpected directory in import file")
+		}
+
+		if skip {
+			continue
+		}
+
+		if header.Name == "export.json" {
+			exportFound = true
+		}
+
+		target := path.Join(p, header.Name)
+		t, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode))
+		if err != nil {
+			return false, 0, fmt.Errorf("failed creating file `%s`: %v", target, err)
+		}
+		defer t.Close()
+		writtenSize, err := io.Copy(t, tr)
+		if err != nil {
+			return false, 0, fmt.Errorf("failed copying file `%s`: %v", target, err)
+		}
+		size += writtenSize
+	}
+	return exportFound, size, nil
+}
+
+func moveCachedSnapshots(names []string, id uint64, p string) ([]string, error) {
```

Comment:
this should identify if we have a snapshots already with the same data using content hashes defined leaving out set id and timestamps